src/hidden.rs

```diff
@@ -20,74 +20,202 @@
 // WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 // USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-//! A wrapper to conceal secrets when output into logs or displayed.
+//! Sometimes you need to handle sensitive data, like a passphrase or key material.
+//! There are pitfalls here that you want to avoid: the data should be zeroized when it goes out of scope, shouldn't be
+//! displayed or output to debug logs, shouldn't be unintentionally copied or moved, and often should have strict type
+//! differentiation to avoid it being misused in an unintended context. This library provides a generic type that can
+//! help.
 
-use std::fmt;
+use std::{any::type_name, fmt};
 
 use serde::{Deserialize, Serialize};
+use zeroize::Zeroize;
 
-/// A simple struct with a single inner value to wrap content of any type.
-#[derive(Copy, Clone, Serialize, Deserialize, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+/// This is a macro that produces a hidden type.
+/// You can use it to generate multiple hidden types that can't be unintentionally mixed up in things like function
+/// definitions.
+///
+/// ```edition2018
+/// # #[macro_use] extern crate tari_utilities;
+/// # use serde::{Deserialize, Serialize};
+/// # use tari_utilities::Hidden;
+/// # use zeroize::Zeroize;
+/// # fn main() {
+/// hidden_type!(MyHiddenType, [u8; 32]);
+/// let example = MyHiddenType::default();
+/// assert_eq!(example.reveal(), &[0u8; 32]);
+/// # }
+/// ```
+#[macro_export]
+macro_rules! hidden_type {
+    ($name:ident, $type:ty) => {
+        /// A hidden type
+        #[derive(Clone, Debug, Default, Deserialize, Eq, Ord, PartialEq, PartialOrd, Serialize, Zeroize)]
+        #[serde(transparent)]
+        pub struct $name
+        where
+            $type: Default + Zeroize,
```

Comment:
nit: These trait bounds aren't strictly necessary. Default is not a bound requirement of Hidden, and Zeroize will be enforced (compile failure) in Hidden anyway.

$type is a concrete type, so except in a few cases, doesn't need trait bounds
```rust
        where
            [u8; 32]: Default + Zeroize,
```

Reply:
The original idea was to simplify instantiation of key types that will be filled in place later. However, this falls down pretty quickly since even types like `[u8; N]` don't implement `Default` for `N > 32`.

Interestingly, we can't even do `Serialize` on such types, which means we might need to drop the `derive` macros on `Hidden` for those traits.